File: test/parsergenerator/firstset_test.scala
package pg.parsergenerator

import org.scalatest.FunSpec
import pg.parsergenerator.firstset.{FirstSet,generateFirst,get,getFromList}
import pg.parsergenerator.symboldiscriminator.genSymbolDiscriminator
import pg.{data, token}
import data.sample_language.{test_empty_language, test_sample_grammar}

class firstset_test extends FunSpec {

  describe("FirstSet test") {
    val first = generateFirst(test_sample_grammar, genSymbolDiscriminator(test_sample_grammar))
    describe("valid one terminal and nonterminal symbol") {
      it("First(S) is {SEMICOLON, SEPARATE, ATOM, ID}") {
        for (symbol <- Array("SEMICOLON", "SEPARATE", "ATOM", "ID")) {
          assert(get(first, "S").contains(symbol))
        }
        assert(get(first, "S").size == 4)
      }
      it("First(E) is {SEMICOLON, SEPARATE, ATOM, ID}") {
        for (symbol <- Array("SEMICOLON", "SEPARATE", "ATOM", "ID")) {
          assert(get(first, "E").contains(symbol))
        }
        assert(get(first, "E").size == 4)
      }
      it("First([E]) is {SEMICOLON, SEPARATE, ATOM, ID}") {
        for (symbol <- Array("SEMICOLON", "SEPARATE", "ATOM", "ID")) {
          assert(getFromList(first, List("E")).contains(symbol))
        }
        assert(getFromList(first, List("E")).size == 4)
      }
      it("First(LIST) is {SEPARATE, ATOM}") {
        for (symbol <- Array("SEPARATE", "ATOM")) {
          assert(get(first, "LIST").contains(symbol))
        }
        assert(get(first, "LIST").size == 2)
      }
      it("First(T) is {ATOM}") {
        assert(get(first, "T").contains("ATOM"))
        assert(get(first, "T").size == 1)
      }
      it("First(HOGE) is {ID}") {
        assert(get(first, "HOGE").contains("ID"))
        assert(get(first, "HOGE").size == 1)
      }
      it("First(ID) is {ID}") {
        assert(get(first, "ID").contains("ID"))
        assert(get(first, "ID").size == 1)
      }
    }
    describe("valid word (multiple terminal or nonterminal symbols)") {
      it("First(LIST ID) is {SEPARATE ATOM ID}") {
        for (symbol <- Array("SEPARATE", "ATOM", "ID")) {
          assert(getFromList(first, List("LIST", "ID")).contains(symbol))
        }
        assert(getFromList(first, List("LIST", "ID")).size == 3)
      }
      it("First(HOGE HOGE) is {ID}") {
        assert(getFromList(first, List("HOGE", "HOGE")).contains("ID"))
        assert(getFromList(first, List("HOGE", "HOGE")).size == 1)
      }
    }
    describe("invalid input (contains neither terminal nor nonterminal symbols)") {
      it("First(FOO) throws error") {
        assertThrows[Error](get(first, "FOO")) //invalid token
      }
      it("First(INVALID) throws error") {
        assertThrows[Error](get(first, "INVALID")) //invalid token
      }
      it("First(INVALID INVALID) throws error") {
        assertThrows[Error](getFromList(first, List("INVALID", "INVALID"))) //invalid token
      }
      it("First(INVALID S) throws error") {
        assertThrows[Error](getFromList(first, List("INVALID", "S"))) //invalid token
      }
      it("First(S INVALID) throws error") {
        assertThrows[Error](getFromList(first, List("S", "INVALID"))) //invalid token
      }
    }
  }

  describe("FirstSet test(empty language)") {
    val first = generateFirst(test_empty_language.grammar, genSymbolDiscriminator(test_empty_language.grammar))
    it("First(S) is {}") {
      assert(get(first, "S").size == 0)
    }
  }
}

File: test/yacc/firstset_test.ml
open OUnit
open Token
open Language
open Utils
open Firstset
open Symboldiscriminator
open Sample_language

let test () =
  let first = generateFirst(test_sample_grammar, genSymbolDiscriminator(test_sample_grammar)) in
  "FirstSet test" >::: [
    "valid one terminal and nonterminal symbol" >::: [
      "First(S) is {SEMICOLON, SEPARATE, ATOM, ID}" >:: begin fun () ->
        ["SEMICOLON"; "SEPARATE"; "ATOM"; "ID"] |> List.iter (fun symbol ->
          assert(S.mem symbol (get(first, "S")))
        );
        assert_equal (S.cardinal(get(first, "S"))) 4
      end;
      "First(E) is {SEMICOLON, SEPARATE, ATOM, ID}" >:: begin fun () ->
        ["SEMICOLON"; "SEPARATE"; "ATOM"; "ID"] |> List.iter (fun symbol ->
          assert(S.mem symbol (get(first, "E")))
        );
        assert_equal (S.cardinal(get(first, "E"))) 4
      end;
      "First([E]) is {SEMICOLON, SEPARATE, ATOM, ID}" >:: begin fun () ->
        ["SEMICOLON"; "SEPARATE"; "ATOM"; "ID"] |> List.iter (fun symbol ->
          assert(S.mem symbol (getFromList(first, ["E"])))
        );
        assert_equal (S.cardinal(getFromList(first, ["E"]))) 4
      end;
      "First(LIST) is {SEPARATE, ATOM}" >:: begin fun () ->
        ["SEPARATE"; "ATOM"] |> List.iter (fun symbol ->
          assert(S.mem symbol (get(first, "LIST")))
        );
        assert_equal (S.cardinal(get(first, "LIST"))) 2
      end;
      
      "First(T) is {ATOM}" >:: begin fun () ->
        assert(S.mem "ATOM" (get(first, "T")));
        assert_equal (S.cardinal(get(first, "T"))) 1
      end;
      "First(HOGE) is {ID}" >:: begin fun () ->
        assert(S.mem "ID" (get(first, "HOGE")));
        assert_equal (S.cardinal(get(first, "HOGE"))) 1
      end;
      "First(ID) is {ID}" >:: begin fun () ->
        assert(S.mem "ID" (get(first, "ID")));
        assert_equal (S.cardinal(get(first, "ID"))) 1
      end;
    ];
    "valid word (multiple terminal or nonterminal symbols)" >::: [
      "First(LIST ID) is {SEPARATE ATOM ID}" >:: begin fun () ->
        ["SEPARATE"; "ATOM"; "ID"] |> List.iter (fun symbol ->
          assert(S.mem symbol (getFromList(first, ["LIST"; "ID"])))
        );
        assert_equal (S.cardinal(getFromList(first, ["LIST"; "ID"]))) 3
      end;
      "First(HOGE HOGE) is {ID}" >:: begin fun () ->
        assert(S.mem "ID" (getFromList(first, ["HOGE"; "HOGE"])));
        assert_equal (S.cardinal(getFromList(first, ["HOGE"; "HOGE"]))) 1
      end;
    ];
    "invalid input (contains neither terminal nor nonterminal symbols)" >::: [
      "First(FOO) throws error" >:: begin fun () ->
        assert_raises (Failure "invalid token found: FOO")
          (fun () -> get(first, "FOO"))
      end;
      "First(INVALID) throws error" >:: begin fun () ->
        assert_raises (Failure "invalid token found: INVALID")
          (fun () -> get(first, "INVALID"))
      end;
      "First(INVALID INVALID) throws error" >:: begin fun () ->
        assert_raises (Failure "invalid token found: INVALID")
          (fun () -> getFromList(first, ["INVALID"; "INVALID"]))
      end;
      "First(INVALID S) throws error" >:: begin fun () ->
        assert_raises (Failure "invalid token found: INVALID")
          (fun () -> getFromList(first, ["INVALID"; "S"]))
      end;
      "First(S INVALID) throws error" >:: begin fun () ->
        assert_raises (Failure "invalid token found: INVALID")
          (fun () -> getFromList(first, ["S"; "INVALID"]))
      end;
    ];
  ]

let test2 () =
  let grammar = test_empty_language.grammar in
  let first = generateFirst(grammar, genSymbolDiscriminator(grammar)) in
  "FirstSet test(empty language)" >::: [
    "First(S) is {}" >:: begin fun () ->
      assert_equal (S.cardinal(get(first, "S"))) 0
    end;
  ]
